File: member1_frontend/src/types/entities.ts
export type EntityType = 
  | 'Person'
  | 'Phone'
  | 'BankAccount'
  | 'Vehicle'
  | 'Location'
  | 'Organization'
  | 'FIR'
  | 'Crime'
  | 'Transaction'
  | 'Communication'
  | 'Evidence';

export interface BaseEntity {
  id: string;
  type: EntityType;
  label: string;
  source: string;
  sourceDocument?: string;
  firstObserved: string;
  lastUpdated: string;
  caseIds: string[];
  evidenceCount: number;
  anomalyIndicators?: string[]; // strictly analytical findings, NOT risk scores
  notes?: string;
}

export interface PersonEntity extends BaseEntity {
  type: 'Person';
  fullName: string;
  aliases: string[];
  dateOfBirth?: string;
  gender?: string;
  nationality?: string;
  nationalIdNumber?: string; // Aadhaar / Passport / Voter ID
  phoneNumbers: string[];
  bankAccounts: string[];
  vehicles: string[];
  addresses: string[];
  organizations: string[];
  associatedFIRs: string[];
  crimesReferenced: string[];
  knownAssociates: { personId: string; name: string; relationType: string; confidence: string }[];
  analyticalSummary: string; // Investigator analytical context
}

export interface PhoneEntity extends BaseEntity {
  type: 'Phone';
  phoneNumber: string;
  imei?: string;
  imsi?: string;
  serviceProvider: string;
  registeredSubscriber: string;
  associatedPersonId?: string;
  cdrCallCount: number;
  frequentTowerLocations: string[];
  lastActiveTower: string;
}

export interface BankAccountEntity extends BaseEntity {
  type: 'BankAccount';
  accountNumber: string;
  bankName: string;
  branch: string;
  ifscCode: string;
  accountHolderName: string;
  accountType: 'Savings' | 'Current' | 'NRE' | 'Escrow';
  associatedPersonId?: string;
  associatedOrgId?: string;
  totalTransactionsLogged: number;
  swiftCode?: string;
}

export interface VehicleEntity extends BaseEntity {
  type: 'Vehicle';
  registrationNumber: string;
  makeModel: string;
  vehicleType: 'Car' | 'Motorcycle' | 'Truck' | 'Container' | 'Vessel';
  registeredOwner: string;
  chassisNumber?: string;
  tollPingsCount: number;
  lastSightedLocation?: string;
}

export interface LocationEntity extends BaseEntity {
  type: 'Location';
  locationName: string;
  address: string;
  city: string;
  state: string;
  latitude: number;
  longitude: number;
  locationCategory: 'Crime Scene' | 'Suspect Residence' | 'Cell Tower' | 'Warehouse' | 'Port / Terminal' | 'Office';
  associatedCrimeIds: string[];
}

export interface OrganizationEntity extends BaseEntity {
  type: 'Organization';
  orgName: string;
  registrationNumber?: string;
  orgType: 'Shell Company' | 'Logistics' | 'Financial Intermediary' | 'Retail' | 'Trust';
  directors: string[];
  registeredAddress: string;
  bankAccountNumbers: string[];
}

export interface FIREntity extends BaseEntity {
  type: 'FIR';
  firNumber: string;
  policeStation: string;
  registrationDate: string;
  sectionsApplied: string[]; // e.g. IPC/BNS sections
  complainantName: string;
  accusedNames: string[];
  investigatingOfficer: string;
  status: 'Under Investigation' | 'Charge Sheet Filed' | 'Closed' | 'Trial Pending';
  briefFacts: string;
}

export interface CrimeEntity extends BaseEntity {
  type: 'Crime';
  crimeId: string;
  crimeCategory: string; // e.g., Narcotics Smuggling, Hawala, Cyber Fraud
  dateOfOccurrence: string;
  locationId: string;
  firNumber: string;
  modusOperandi: string;
  seizedMaterials: string[];
}

export interface TransactionEntity extends BaseEntity {
  type: 'Transaction';
  transactionId: string;
  sourceAccount: string;
  destinationAccount: string;
  sourceHolder: string;
  destinationHolder: string;
  amountINR: number;
  timestamp: string;
  channel: 'NEFT' | 'RTGS' | 'IMPS' | 'SWIFT' | 'Hawala Ledger' | 'Cash Deposit';
  anomalyNote?: string;
}

export interface CommunicationEntity extends BaseEntity {
  type: 'Communication';
  callId: string;
  callerNumber: string;
  receiverNumber: string;
  callerName?: string;
  receiverName?: string;
  callType: 'Voice' | 'SMS' | 'VoIP / Encrypted' | 'WhatsApp Call';
  durationSeconds: number;
  timestamp: string;
  originTower: string;
  destinationTower?: string;
}

export interface EvidenceEntity extends BaseEntity {
  type: 'Evidence';
  evidenceId: string;
  title: string;
  category: 'Digital Forensic Image' | 'CDR Dump' | 'Bank Statement' | 'Seized Physical Item' | 'CCTV Footage' | 'FIR Copy';
  originalHashSHA256: string;
  currentHashSHA256: string;
  isIntegrityVerified: boolean;
  seizureDate: string;
  seizingOfficer: string;
  chainOfCustodyCurrentHolder: string;
  bsaSection63CertificateId?: string;
  fileSizeBytes: number;
  mimeType: string;
  downloadUrl?: string;
}

export type AnyEntity = 
  | PersonEntity 
  | PhoneEntity 
  | BankAccountEntity 
  | VehicleEntity 
  | LocationEntity 
  | OrganizationEntity 
  | FIREntity 
  | CrimeEntity 
  | TransactionEntity 
  | CommunicationEntity 
  | EvidenceEntity;

File: member2_backend/app/schemas/entities.py
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

class BaseEntity(BaseModel):
    id: str
    entityType: str
    canonicalName: str
    source: str = 'M3_GRAPH_DATA'
    caseId: Optional[str] = None
    evidenceId: Optional[str] = None
    confidence: Optional[float] = Field(default=0.95, ge=0.0, le=1.0)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    createdAt: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())

class PersonEntity(BaseEntity):
    entityType: str = 'Person'
    fullName: str
    aliases: List[str] = Field(default_factory=list)
    dateOfBirth: Optional[str] = None
    nationalId: Optional[str] = None
    passportNumber: Optional[str] = None
    address: Optional[str] = None
    associatedPhones: List[str] = Field(default_factory=list)
    associatedAccounts: List[str] = Field(default_factory=list)

class PhoneEntity(BaseEntity):
    entityType: str = 'Phone'
    phoneNumber: str
    imei: Optional[str] = None
    carrier: Optional[str] = None
    subscriberName: Optional[str] = None
    callCount: Optional[int] = 0
    smsCount: Optional[int] = 0

class BankAccountEntity(BaseEntity):
    entityType: str = 'BankAccount'
    accountNumber: str
    bankName: str
    branch: Optional[str] = None
    ifscCode: Optional[str] = None
    accountHolder: str
    totalCredits: Optional[float] = 0.0
    totalDebits: Optional[float] = 0.0

class VehicleEntity(BaseEntity):
    entityType: str = 'Vehicle'
    registrationNumber: str
    make: Optional[str] = None
    model: Optional[str] = None
    color: Optional[str] = None
    chassisNumber: Optional[str] = None
    engineNumber: Optional[str] = None
    ownerName: Optional[str] = None

class LocationEntity(BaseEntity):
    entityType: str = 'Location'
    locationName: str
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    country: str = 'India'
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    locationType: Optional[str] = 'Transit Point'

class OrganizationEntity(BaseEntity):
    entityType: str = 'Organization'
    orgName: str
    registrationNumber: Optional[str] = None
    orgType: Optional[str] = 'Shell Corporation'
    directors: List[str] = Field(default_factory=list)
    registeredAddress: Optional[str] = None

class FIREntity(BaseEntity):
    entityType: str = 'FIR'
    firNumber: str
    policeStation: str
    filingDate: str
    actsSections: List[str] = Field(default_factory=list)
    complainant: Optional[str] = None
    accusedPersons: List[str] = Field(default_factory=list)
    incidentSummary: str

class CrimeEntity(BaseEntity):
    entityType: str = 'Crime'
    crimeCode: str
    crimeCategory: str
    description: str
    incidentDate: str
    location: Optional[str] = None
    status: str = 'Under Investigation'

class TransactionEntity(BaseEntity):
    entityType: str = 'Transaction'
    transactionId: str
    sourceAccount: str
    destinationAccount: str
    amount: float
    currency: str = 'INR'
    transactionDate: str
    paymentChannel: str = 'NEFT/RTGS'
    referenceNumber: Optional[str] = None

class CommunicationEntity(BaseEntity):
    entityType: str = 'Communication'
    commId: str
    callerPhone: str
    receiverPhone: str
    commType: str = 'Voice Call'
    timestamp: str
    durationSeconds: Optional[int] = 0
    towerLocation: Optional[str] = None

class EvidenceEntity(BaseEntity):
    entityType: str = 'Evidence'
    evidenceNumber: str
    evidenceType: str
    description: str
    collectedDate: str
    collectedBy: str
    storageLocation: str
    sha256Hash: str
    bsaSection65BCertificateId: Optional[str] = None

class AliasEntity(BaseModel):
    aliasId: str
    primaryPersonId: str
    aliasName: str
    context: Optional[str] = None
    source: str = 'M4_NLP_EXTRACTION'

class EntityFilter(BaseModel):
    entityType: Optional[str] = None
    query: Optional[str] = None
    caseId: Optional[str] = None
    location: Optional[str] = None
    fromDate: Optional[str] = None
    toDate: Optional[str] = None
    source: Optional[str] = None
